crates/host/src/wasmbus/lattice.rs

```diff
@@ -0,0 +1,2954 @@
+                                    Err(error) => {
+                                        error!("failed to watch lattice data bucket: {error}");
+                                    }
+                                    Ok(entry) => lattice.process_entry(entry, true).await,
+                                }
+                            }
+                        }
+                    })
+                    .await;
+                let deadline = { *lattice.stop_rx.borrow() };
+                lattice.stop_tx.send_replace(deadline);
+                if data_watch.is_aborted() {
+                    info!(lattice = name, "data watch task gracefully stopped");
+                } else {
+                    error!(lattice = name, "data watch task unexpectedly stopped");
+                }
+                Ok(())
+            }
+        });
+
+        let queue = Queue::new_lattice(
+            &ctl_client.clone(),
+            &config.ctl_topic_prefix.clone(),
+            &name,
+            &host_key,
+        )
+        .await
+        .context("failed to initialize queue")?;
+
+        let _queue = spawn({
+            let lattice = Arc::clone(&lattice);
+            async move {
+                let mut queue = Abortable::new(queue, queue_abort_reg);
+                queue
+                    .by_ref()
+                    .for_each_concurrent(None, {
+                        let lattice = Arc::clone(&lattice);
+                        move |msg| {
+                            let lattice = Arc::clone(&lattice);
+                            async move { lattice.handle_ctl_message(msg).await }
+                        }
+                    })
+                    .await;
+                let deadline = { *lattice.stop_rx.borrow() };
+                lattice.stop_tx.send_replace(deadline);
+                if queue.is_aborted() {
+                    info!("control interface queue task gracefully stopped");
+                } else {
+                    error!("control interface queue task unexpectedly stopped");
+                }
+            }
+        });
+
+        // Shutdown when requested
+        spawn({
+            let mut stop = lattice.stop_rx.clone();
+            let lattice = Arc::clone(&lattice);
+            let name = lattice.name.clone();
+            async move {
+                if let Err(e) = stop.changed().await {
+                    error!(lattice =% name.clone(), "failed to wait for stop: {e}");
+                }
+                if let Err(err) = lattice.shutdown().await {
+                    error!(lattice =% name, "failed to shutdown lattice: {err}");
+                };
+                info!(lattice =% name, "lattice stopped");
+            }
+        });
+
+        // TODO this feels like a host-level responsibility and not a lattice thing. The problem of
+        // course is that we need to fix the control interface so that a host heartbeat contains
+        // information from all lattices.
```

Comment:
```suggestion
        // TODO(#3120) this feels like a host-level responsibility and not a lattice thing. The problem of
        // course is that we need to fix the control interface so that a host heartbeat contains
        // information from all lattices.
```